### ISSUE TITLE
Build on FreeBSD fails because libresolv is missing.

### ISSUE THREAD

**Issue by gwright83** [bictorv/chaosnet-bridge#33]
The build on FreeBSD fails because libresolv does not exist on that platform.  (Name resolution functions are in the standard C library.)

Pull request #32 patches the Makefile to correct this.

**Comment by bictorv** [comment]
Are the standard C library functions threadsafe? Or would FreeBSD fit under the OpenBSD settings in the Makefile? CC @ams.

**Comment by gwright83** [comment]
In FreeBSD the standard libc functions are supposed to be thread-safe when linked with -lthr.  In that case, thread-safe versions are linked in.

FreeBSD does have libbind in ports, so it would be possible to require it and link just as with OpenBSD.  However, on OpenBSD, the libbind port comes with the warning:

> This package provides the ISC libbind DNS resolver.
>
>This port is not for general purpose use and has known issues; it is
>intended *only* for special cases where software relies on resolver
>internals and cannot be adapted to the standard resolver in libc
>or use an alternative library. There are symbol conflicts; expect
>crashes and corruption in some cases.

Does chaosnet-bridge require access to the internals of the resolver?  Or were there problems in the past with the resolver functions in OpenBSD's libc?

**Comment by bictorv** [comment]
cbridge uses the resolver to find Chaosnet class data. It also parses the results, and in the Chaosnet DNS service "manually" truncates them (or tries to) to fit in a Chaos packet. But if you didn't get compiler warnings, I guess the chances are good that the FreeBSD libc version works - but make sure to add -lthr in Makefile.

I don't recall the details of the OpenBSD case, but @ams possibly does? Should -lthr be added there too, by the way? Are there other libraries which need -lthr to be threadsafe?

**Comment by gwright83** [comment]
On FreeBSD, libthr is just a symlink to libphtread:
> root@hardtack:~/tmp/cbridge-master # ls -l /usr/lib/libpthread*
>lrwxr-xr-x  1 root  wheel   8 Aug 21  2022 /usr/lib/libpthread.a -> libthr.a
>lrwxr-xr-x  1 root  wheel   9 Aug 21  2022 /usr/lib/libpthread.so -> libthr.so
>lrwxr-xr-x  1 root  wheel  10 Aug 21  2022 /usr/lib/libpthread_p.a -> libthr_p.a
>root@hardtack:~/tmp/cbridge-master #
so it is already linked in by the `-lpthread` directive in the Makefile.

`libthr` does not exist on OpenBSD.

The linking directive should probably be changed from `-lpthread` to `-pthread`.  The former simply links in the pthread library; the latter directs the linker to override the weak references in other libraries with the appropriate threaded versions.

**Comment by gwright83** [comment]
The build succeeds on OpenBSD 7.3 using the -pthread compiler option and with libbind installed.  There are, as usual for OpenBSD, more warning messages:

`root@mailserver1:~/cbridge-master # gmake clean
rm -f cbridge hostat finger cbridge.o contacts.o usockets.o chtls.o chudp.o debug.o chether.o dns.o chip.o ncp.o pkqueue.o
root@mailserver1:~/cbridge-master # gmake
cc -c -I/usr/local/include/bind -g -o cbridge.o cbridge.c
cbridge.c:378:4: warning: format specifies type 'long' but the argument has type 'long long' [-Wformat]
                 he[i].hst_last_seen > 0 ? time(NULL) - he[i].hst_last_seen : 0,
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cbridge.c:429:9: warning: format specifies type 'long' but the argument has type 'long long' [-Wformat]
               rttbl_host[i].rt_cost_updated > 0 ? time(NULL) - rttbl_host[i].rt_cost_updated : 0);
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cbridge.c:449:7: warning: format specifies type 'long' but the argument has type 'long long' [-Wformat]
             rttbl_net[i].rt_cost_updated > 0 ? time(NULL) - rttbl_net[i].rt_cost_updated : 0);
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cbridge.c:1736:8: warning: format specifies type 'unsigned int' but the argument has type 'unsigned long' [-Wformat]
              addr);
              ^~~~
4 warnings generated.
cc -c -I/usr/local/include/bind -g -o contacts.o contacts.c
contacts.c:223:6: warning: format specifies type 'unsigned long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
            t, data[0], data[1], data[2], data[3]);
            ^
1 warning generated.
cc -c -I/usr/local/include/bind -g -o usockets.o usockets.c
cc -c -I/usr/local/include/bind -g -o chtls.o chtls.c
chtls.c:1439:25: warning: incompatible pointer types passing 'u_short (*)[4]' to parameter of type 'u_short *' (aka 'unsigned short *') [-Wincompatible-pointer-types]
         is_in_mux_list(dchad, &tlsdest[i].tls_muxed)
                               ^~~~~~~~~~~~~~~~~~~~~
chtls.c:1404:39: note: passing argument to parameter 'list' here
is_in_mux_list(u_short addr, u_short *list)
                                      ^
1 warning generated.
cc -c -I/usr/local/include/bind -g -o chudp.o chudp.c
cc -c -I/usr/local/include/bind -g -o debug.o debug.c
cc -c -I/usr/local/include/bind -g -o chether.o chether.c
chether.c:1098:7: warning: format specifies type 'unsigned long' but the argument has type 'long long' [-Wformat]
             (time(NULL) - charp_list[i].charp_age),
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
chether.c:1118:16: warning: format specifies type 'unsigned long' but the argument has type 'long long' [-Wformat]
                             daddr, (time(NULL) - charp_list[i].charp_age));
                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 warnings generated.
cc -c -I/usr/local/include/bind -g -o dns.o dns.c
cc -c -I/usr/local/include/bind -g -o chip.o chip.c
cc -c -I/usr/local/include/bind -g -o ncp.o ncp.c
ncp.c:516:26: warning: format specifies type 'long' but the argument has type 'long long' [-Wformat]
         conn_state_name(conn), now - conn->conn_created, conn->rfc_timeout,
                                ^~~~~~~~~~~~~~~~~~~~~~~~
ncp.c:527:5: warning: format specifies type 'long' but the argument has type 'long long' [-Wformat]
           cs->time_last_received > 0 ? now - cs->time_last_received : -1,
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ncp.c:528:5: warning: format specifies type 'long' but the argument has type 'long long' [-Wformat]
           cs->time_last_probed > 0 ? now - cs->time_last_probed : -1);
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ncp.c:1421:44: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
           timedout != 0 ? "timed out" : "acked", tn.tv_sec, tn.tv_usec/1000);
                                                  ^~~~~~~~~
ncp.c:1421:55: warning: format specifies type 'int' but the argument has type 'long' [-Wformat]
           timedout != 0 ? "timed out" : "acked", tn.tv_sec, tn.tv_usec/1000);
                                                             ^~~~~~~~~~~~~~~
ncp.c:2557:5: warning: format specifies type 'long' but the argument has type 'long long' [-Wformat]
                                time(NULL) - conn->conn_created, conn->rfc_timeout);
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ncp.c:2662:49: warning: format specifies type 'long' but the argument has type 'long long' [-Wformat]
                            conn, conn->conn_lhost, conn->conn_lidx, time(NULL) - conn->conn_created);
                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ncp.c:2671:6: warning: format specifies type 'long' but the argument has type 'long long' [-Wformat]
                          now.tv_sec - cs->time_last_received);
                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ncp.c:2687:8: warning: format specifies type 'long' but the argument has type 'long long' [-Wformat]
                            now.tv_sec - cs->time_last_received);
                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ncp.c:2778:5: warning: format specifies type 'long' but the argument has type 'long long' [-Wformat]
                                (now.tv_sec*1000 + now.tv_nsec/1000000) - (lastsent.tv_sec*1000 + lastsent.tv_nsec/1000000));
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ncp.c:3262:34: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
                  strerror(errno), sval, sock, timeout.tv_sec, timeout.tv_usec);
                                               ^~~~~~~~~~~~~~
ncp.c:3262:50: warning: format specifies type 'int' but the argument has type 'suseconds_t' (aka 'long') [-Wformat]
                  strerror(errno), sval, sock, timeout.tv_sec, timeout.tv_usec);
                                                               ^~~~~~~~~~~~~~~
ncp.c:3857:64: warning: format specifies type 'long' but the argument has type 'long long' [-Wformat]
                              conn->conn_lhost, conn->conn_lidx, conn->rfc_timeout, time(NULL) - conn->conn_created);
                                                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
13 warnings generated.
cc -c -I/usr/local/include/bind -g -o pkqueue.o pkqueue.c
cc -I/usr/local/include/bind -g -L/usr/local/lib/libbind -o cbridge cbridge.o contacts.o usockets.o chtls.o chudp.o debug.o chether.o dns.o chip.o ncp.o pkqueue.o -pthread -lssl -lcrypto -lbind
cbridge.c:2123(cbridge.o:(main)): warning: strcpy() is almost always misused, please use strlcpy()
cbridge.c:2163(cbridge.o:(main)): warning: sprintf() is often misused, please use snprintf()
ncp.c:124(ncp.o:(parse_ncp_config_line)): warning: strcat() is almost always misused, please use strlcat()
ncp.c:313(ncp.o:(make_u_short_random)): warning: random() may return deterministic values, is that what you want?
ncp.c:1727(ncp.o:(user_socket_los)): warning: vsprintf() is often misused, please use vsnprintf()
cc -I/usr/local/include/bind -g -o hostat hostat.c
In file included from hostat.c:32:
In file included from ./cbridge-chaos.h:1:
/usr/local/include/bind/arpa/inet.h:93:43: warning: declaration of 'struct in_addr' will not be visible outside of this function [-Wvisibility]
int              inet_aton __P((const char *, struct in_addr *));
                                                     ^
/usr/local/include/bind/arpa/inet.h:94:39: warning: declaration of 'struct in_addr' will not be visible outside of this function [-Wvisibility]
unsigned long    inet_lnaof __P((struct in_addr));
                                        ^
hostat.c:231:50: warning: format specifies type 'long' but the argument has type 'time_t' (aka 'long long') [-Wformat]
    printf("Unexpected time value %ld <= %ld\n", t, 2208988800UL);
                                  ~~~            ^
                                  %lld
3 warnings generated.
hostat.c:87(/tmp/hostat-9c02db.o:(ch_char)): warning: sprintf() is often misused, please use snprintf()
cc -I/usr/local/include/bind -g -o finger finger.c
In file included from finger.c:30:
In file included from ./cbridge-chaos.h:1:
/usr/local/include/bind/arpa/inet.h:93:43: warning: declaration of 'struct in_addr' will not be visible outside of this function [-Wvisibility]
int              inet_aton __P((const char *, struct in_addr *));
                                                     ^
/usr/local/include/bind/arpa/inet.h:94:39: warning: declaration of 'struct in_addr' will not be visible outside of this function [-Wvisibility]
unsigned long    inet_lnaof __P((struct in_addr));
                                        ^
2 warnings generated.
finger.c:97(/tmp/finger-a63124.o:(main)): warning: strcat() is almost always misused, please use strlcat()
finger.c:52(/tmp/finger-a63124.o:(connect_to_named_socket)): warning: sprintf() is often misused, please use snprintf()
root@mailserver1:~/cbridge-master #`

I did verify that on OpenBSD 7.3, libbind is still required.  If omitted, the build fails.

I will produce an updated pull request with the current (as tested) version of the Makefile.

**Comment by bictorv** [comment]
Thanks for all the help! Resolved by #32
